ls1intum/Artemis
# [Feature] Grading System - Course and Exam Scores

### Checklist

### Motivation and Context


This PR provides integration for the grading system on the course and exam scores pages. The instructor would be able to view more detailed information if a grading scale for the given course/exam has been set.

### Description

I've mainly modified the course and exam scores components by first checking if a grading scale exists for the course/exam and then calculating additional properties that get displayed in the statistics tables/graphs based on that.

### Steps for Testing


1 Log in to Artemis as an instructor

2.1 Navigate to Course Administration -> Course with some exercise submissions
3.1 Open the scores page to see what it looks like for reference
4.1 Create a Grade Key of "Grade" type for the course and check the scores page again
a) there should be a grades column that has a valid mapping to the grade key based on the overall percentage
b) test if the exported file has additional grades information
5.1 Create a Grade Key of "Bonus" type for that course and check the scores page again
a) there should be a bonus points column  that has a valid mapping to the grade key based on the overall percentage
b) test if the exported file has additional bonus points information

2.2. Navigate to Course Administration -> Exams and find an Exam with existing student submissions (alternatively create one and do multiple student submissions with different grading outcomes)
3.2 Make sure the exam has no grade key set and check the scores page to see what it looks like if that's the case
4.2 Create a Grade Key of "Grade" type and check the scores page again
a) additional statistics should be displayed like average and median grade, statistics for students who have passed the exam
b) the graph should display the grade key mapping instead of the default one and the scores should again be mapped to the correct grades
c) exercise groups table should have an average grade column with the correct grade
d) student table should have an "overall grade" and a "passed" column with the correct values
e) test exporting
5.2 Now change the Grade Key to "Bonus" type
a) check that the column names are different (bonus points instead of grades)
b) the "Passing" columns in both the overview and student table shouldn't be present
c) the graph should map to the bonus points grade key
d) test exporting


6 Delete the Grade Key and check that the scores page is again presenting the default values
7 Make sure that no TypeErrors (or any Errors) occur in the console when opening the page with or without a grade key set.
### Test Coverage



grading-system.service.ts: 100%
course-scores.component.ts: 91.05%
exam-scores.component.ts: 86.2%

### Screenshots


![Screenshot 2021-05-19 104128](https://user-images.githubusercontent.com/63168603/118778601-f0d29280-b892-11eb-93ec-0304e5748584.png)
![Screenshot 2021-05-19 105629](https://user-images.githubusercontent.com/63168603/118778605-f16b2900-b892-11eb-9860-5371a0c2656e.png)
![Screenshot 2021-05-19 105642](https://user-images.githubusercontent.com/63168603/118778607-f203bf80-b892-11eb-9182-cf63a9e79903.png)
![Screenshot 2021-05-19 111031](https://user-images.githubusercontent.com/63168603/118778608-f203bf80-b892-11eb-8f08-a3650c818279.png)
![Screenshot 2021-05-19 111047](https://user-images.githubusercontent.com/63168603/118778610-f29c5600-b892-11eb-8e7e-67d3b5071506.png)
![Screenshot 2021-05-19 104027](https://user-images.githubusercontent.com/63168603/118778611-f29c5600-b892-11eb-8c85-13768d75e196.png)
